doc/manual/merging.md

```diff
@@ -0,0 +1,522 @@
+  succeed, although a closed contract `{subfield | Str}` is attached to `foo`,
+  and the final result would have an additional field `other_subfield` ?
+
+Nickel chooses to answer **no** to both. In general, when a contract is attached
+to a field `foo`, merging ensures that whatever is this field merged with,
+including being dropped in favor of another value, the final value for `foo` has
+to respect the contract as well or the evaluation will fail accordingly.
+
+### Specification
+
+For two operands with one field each, which the same on both side, with respective
+contracts `Left1, .., Leftn` and `Right1, .., Rightk` attached:
+
+```nickel
+left = {
+  common | Left1
+         | ..
+         | Leftn
+}
+```
+
+And
+
+```nickel
+right = {
+  common | Right1
+         | ..
+         | Rightk
+}
+```
+
+Then the `common` field of `left & right` will be checked against `Left1, ..,
+Leftn, Right1, .., Rightk`. Here, we ignore the case of type annotations such as
+`common: LeftType` that can just be considered as an additional contract
+`Left0`.
+
+### Example
+
+```nickel
+let Port
+  | doc "A valid port number"
+  = contract.from_predicate (fun value =>
+    builtin.is_num value &&
+    value % 1 == 0 &&
+    value >= 0 &&
+    value <= 65535) in
+let GreaterThan
+  | doc "A number greater than the paramater"
+  = fun x => contract.from_predicate (fun value => value > x) in
+
+{
+    port | GreaterThan 1024
+         | default = 8080,
+} & {
+    port | Port = 80,
+}
+```
+
+This fails at evaluation:
+
+```
+error: contract broken by a value.
+[..]
+   ┌─ repl-input-1:16:19
+   │
+16 │     port | Port = 80,
+   │                   ^^ applied to this expression
+
+note:
+   ┌─ repl-input-1:13:12
+   │
+13 │     port | GreaterThan 1024
+   │            ^^^^^^^^^^^^^^^^ bound here
+```
+
+
+## Documentation
+
+Documentation is attached via the `doc` keyword. Documentation is propagated
+during merging. For example, querying `foo` by using the command `nickel -f
+config.ncl query foo` on:
+
+```nickel
+// config.ncl
+{
+  foo | doc "Some documentation"
+      | = {}
+} & {
+  foo.field = null,
+}
+```
+
+Will print `"Some documentation"` as expected. If both sides have documentation, the behavior
+is unspecified, as merging two distinct blobs of text doesn't always make sense
+in general. Currently, Nickel will randomly keeps one of the two in practice.
+
+## Recursive overriding
+
+We've seen in the section on default values that they are useful to override
+(update) a single field with a different value. The combo of merging and default
+values can do more. In Nickel, records are recursive by default, in order to
+express easily dependencies between the different fields of the configuration.
+Concretely, you can refer to other fields of a record from within this record:
+
+```nickel
+let base_config = {
+  version | default = "20.09",
+  input.url | default = url = "nixpkgs/nixos-%{version}",
```

Comment:
Is that really syntactically correct?

Reply:
Nope, it's inconsistent consecutive editions, good catch.